<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>pybedtools.bedtool.BedTool.genome_coverage &mdash; pybedtools 0.7.0 documentation</title>
    
    <link rel="stylesheet" href="../_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.7.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="pybedtools 0.7.0 documentation" href="../index.html" />
    <link rel="up" title="pybedtools Reference" href="../autodoc_source.html" />
    <link rel="next" title="pybedtools.bedtool.BedTool.merge" href="pybedtools.bedtool.BedTool.merge.html" />
    <link rel="prev" title="pybedtools.bedtool.BedTool.map" href="pybedtools.bedtool.BedTool.map.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pybedtools.bedtool.BedTool.merge.html" title="pybedtools.bedtool.BedTool.merge"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="pybedtools.bedtool.BedTool.map.html" title="pybedtools.bedtool.BedTool.map"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">pybedtools 0.7.0 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../autodoc_source.html" accesskey="U"><code class="docutils literal"><span class="pre">pybedtools</span></code> Reference</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="pybedtools-bedtool-bedtool-genome-coverage">
<h1>pybedtools.bedtool.BedTool.genome_coverage<a class="headerlink" href="#pybedtools-bedtool-bedtool-genome-coverage" title="Permalink to this headline">¶</a></h1>
<dl class="method">
<dt id="pybedtools.bedtool.BedTool.genome_coverage">
<code class="descclassname">BedTool.</code><code class="descname">genome_coverage</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pybedtools/bedtool.html#BedTool.genome_coverage"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pybedtools.bedtool.BedTool.genome_coverage" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div><p>Wraps <code class="docutils literal"><span class="pre">bedtools</span> <span class="pre">genomecov</span></code>.</p>
<p>Example usage:</p>
<p>BAM file input does not require a genome:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">example_bedtool</span><span class="p">(</span><span class="s">&#39;x.bam&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">genome_coverage</span><span class="p">(</span><span class="n">bg</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span> 
<span class="go">chr2L   9329    9365    1</span>
<span class="go">chr2L   10212   10248   1</span>
<span class="go">chr2L   10255   10291   1</span>
</pre></div>
</div>
<p>Other input does require a genome:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">example_bedtool</span><span class="p">(</span><span class="s">&#39;x.bed&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">genome_coverage</span><span class="p">(</span><span class="n">bg</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="s">&#39;dm3&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span> 
<span class="go">chr2L   9329    9365    1</span>
<span class="go">chr2L   10212   10248   1</span>
<span class="go">chr2L   10255   10291   1</span>
</pre></div>
</div>
</div></blockquote>
<p>For convenience, the file or stream this BedTool points to
is implicitly passed as the <code class="docutils literal"><span class="pre">-i</span></code> argument to <code class="docutils literal"><span class="pre">genomeCoverageBed</span></code></p>
<p>There are two alternatives for supplying a genome.  Use
<code class="docutils literal"><span class="pre">g=&quot;genome.filename&quot;</span></code> if you have a genome's chrom sizes
saved as a file. This is the what BEDTools expects when
using it from the command line. Alternatively, use the
<code class="docutils literal"><span class="pre">genome=&quot;assembly.name&quot;</span></code> (for example, <code class="docutils literal"><span class="pre">genome=&quot;hg19&quot;</span></code>) to
use chrom sizes for that assembly without having to manage
a separate file.  The <code class="docutils literal"><span class="pre">genome</span></code> argument triggers a call
<code class="docutils literal"><span class="pre">pybedtools.chromsizes</span></code>, so see that method for more
details.</p>
<p><strong>Original BEDTools help:</strong>:</p>
<div class="highlight-python"><div class="highlight"><pre>Tool:    bedtools genomecov (aka genomeCoverageBed)
Version: v2.25.0
Summary: Compute the coverage of a feature file among a genome.

Usage: bedtools genomecov [OPTIONS] -i &lt;bed/gff/vcf&gt; -g &lt;genome&gt;

Options: 
        -ibam           The input file is in BAM format.
                        Note: BAM **must** be sorted by position

        -d              Report the depth at each genome position (with one-based coordinates).
                        Default behavior is to report a histogram.

        -dz             Report the depth at each genome position (with zero-based coordinates).
                        Reports only non-zero positions.
                        Default behavior is to report a histogram.

        -bg             Report depth in BedGraph format. For details, see:
                        genome.ucsc.edu/goldenPath/help/bedgraph.html

        -bga            Report depth in BedGraph format, as above (-bg).
                        However with this option, regions with zero 
                        coverage are also reported. This allows one to
                        quickly extract all regions of a genome with 0 
                        coverage by applying: &quot;grep -w 0$&quot; to the output.

        -split          Treat &quot;split&quot; BAM or BED12 entries as distinct BED intervals.
                        when computing coverage.
                        For BAM files, this uses the CIGAR &quot;N&quot; and &quot;D&quot; operations 
                        to infer the blocks for computing coverage.
                        For BED12 files, this uses the BlockCount, BlockStarts, and BlockEnds
                        fields (i.e., columns 10,11,12).

        -strand         Calculate coverage of intervals from a specific strand.
                        With BED files, requires at least 6 columns (strand is column 6). 
                        - (STRING): can be + or -

        -5              Calculate coverage of 5&quot; positions (instead of entire interval).

        -3              Calculate coverage of 3&quot; positions (instead of entire interval).

        -max            Combine all positions with a depth &gt;= max into
                        a single bin in the histogram. Irrelevant
                        for -d and -bedGraph
                        - (INTEGER)

        -scale          Scale the coverage by a constant factor.
                        Each coverage value is multiplied by this factor before being reported.
                        Useful for normalizing coverage by, e.g., reads per million (RPM).
                        - Default is 1.0; i.e., unscaled.
                        - (FLOAT)

        -trackline      Adds a UCSC/Genome-Browser track line definition in the first line of the output.
                        - See here for more details about track line definition:
                              http://genome.ucsc.edu/goldenPath/help/bedgraph.html
                        - NOTE: When adding a trackline definition, the output BedGraph can be easily
                              uploaded to the Genome Browser as a custom track,
                              BUT CAN NOT be converted into a BigWig file (w/o removing the first line).

        -trackopts      Writes additional track line definition parameters in the first line.
                        - Example:
                           -trackopts &#39;name=&quot;My Track&quot; visibility=2 color=255,30,30&#39;
                           Note the use of single-quotes if you have spaces in your parameters.
                        - (TEXT)

Notes: 
        (1) The genome file should tab delimited and structured as follows:
         &lt;chromName&gt;&lt;TAB&gt;&lt;chromSize&gt;

        For example, Human (hg19):
        chr1    249250621
        chr2    243199373
        ...
        chr18**gl000207**random 4262

        (2) The input BED (-i) file must be grouped by chromosome.
         A simple &quot;sort -k 1,1 &lt;BED&gt; &gt; &lt;BED&gt;.sorted&quot; will suffice.

        (3) The input BAM (-ibam) file must be sorted by position.
         A &quot;samtools sort &lt;BAM&gt;&quot; should suffice.

Tips: 
        One can use the UCSC Genome Browser&#39;s MySQL database to extract
        chromosome sizes. For example, H. sapiens:

        mysql --user=genome --host=genome-mysql.cse.ucsc.edu -A -e \
        &quot;select chrom, size from hg19.chromInfo&quot; &gt; hg19.genome
</pre></div>
</div>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="pybedtools.bedtool.BedTool.map.html"
                        title="previous chapter">pybedtools.bedtool.BedTool.map</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="pybedtools.bedtool.BedTool.merge.html"
                        title="next chapter">pybedtools.bedtool.BedTool.merge</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/autodocs/pybedtools.bedtool.BedTool.genome_coverage.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>

    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pybedtools.bedtool.BedTool.merge.html" title="pybedtools.bedtool.BedTool.merge"
             >next</a> |</li>
        <li class="right" >
          <a href="pybedtools.bedtool.BedTool.map.html" title="pybedtools.bedtool.BedTool.map"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">pybedtools 0.7.0 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../autodoc_source.html" ><code class="docutils literal"><span class="pre">pybedtools</span></code> Reference</a> &raquo;</li> 
      </ul>
    </div>
<!-- Piwik --> 
<script type="text/javascript">
var pkBaseURL = (("https:" == document.location.protocol) ? "https://piwik.genomicnorth.com/piwik/" : "http://piwik.genomicnorth.com/piwik/");
document.write(unescape("%3Cscript src='" + pkBaseURL + "piwik.js' type='text/javascript'%3E%3C/script%3E"));
</script><script type="text/javascript">
try {
var piwikTracker = Piwik.getTracker(pkBaseURL + "piwik.php", 1);
piwikTracker.trackPageView();
piwikTracker.enableLinkTracking();
} catch( err ) {}
</script><noscript><p><img src="http://piwik.genomicnorth.com/piwik/piwik.php?idsite=1" style="border:0" alt="" /></p></noscript>
<!-- End Piwik Tracking Code -->


    <div class="footer" role="contentinfo">
        &copy; Copyright 2010, Ryan Dale.
      Last updated on Sep 03, 2015.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>