<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>pybedtools.contrib.classifier &#8212; pybedtools 0.7.9 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '0.7.9',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for pybedtools.contrib.classifier</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pybedtools</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>


<span class="k">class</span> <span class="nc">BasePairClassifier</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bed</span><span class="p">,</span> <span class="n">annotations</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">sample_name</span><span class="o">=</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span>
                 <span class="n">names</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;split_&#39;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Classifies files using bedtools multiinter.</span>

<span class="sd">        The results from this class split up the bp in `bed` in to classes as</span>
<span class="sd">        annotated in `annotations`.  Note that this is different from counting</span>
<span class="sd">        the number of features in `bed` that fall within `annotations`; for</span>
<span class="sd">        this latter functionality, see the FeatureClassifier class in this same</span>
<span class="sd">        module.</span>

<span class="sd">        `bed` must be a BED/VCF/GFF/GTF BedTool object or filename -- not</span>
<span class="sd">        a BAM.  This is because `bedtools multiinter`, the program called here,</span>
<span class="sd">        does not accept BAM as input. If you want to use reads from a BAM file,</span>
<span class="sd">        first convert to bed, i.e.,::</span>

<span class="sd">            bed = pybedtools.BedTool(&#39;reads.bam&#39;).bam_to_bed(split=True).fn</span>

<span class="sd">        `annotations` will be used to classify the features in `bed`.</span>

<span class="sd">            If `annotations` is a list, then use those annotations directly.</span>

<span class="sd">            If `annotations` is a BedTool or string, then split the annotations</span>
<span class="sd">            into multiple files based on the featuretype (assumes GFF/GTF</span>
<span class="sd">            format).  Each of these new filenames will get the `prefix`</span>
<span class="sd">            supplied.</span>

<span class="sd">        `sample_name` is the name of the sample. It is used internally, but it</span>
<span class="sd">        needs to be a unique name that is not the name of a featuretype in the</span>
<span class="sd">        annotations file (i.e., sample_name=&quot;exon&quot; would not be a good choice).</span>

<span class="sd">        `names` is an optional list of names that correspond to `annotations`</span>
<span class="sd">        if it is a list, OR a function that maps items in `annotations` to</span>
<span class="sd">        a featuretype name -- e.g.::</span>

<span class="sd">            names=lambda x: x.replace(&#39;.gff&#39;, &#39;&#39;)</span>

<span class="sd">        `genome` is the string assembly name, or dictionary of {chrom: (start,</span>
<span class="sd">        stop)} coordinates for each chromosome.  It is used to determine the</span>
<span class="sd">        &quot;unannotated&quot; space of the genome.</span>

<span class="sd">        After running the `classify()` method, the results BedTool is available</span>
<span class="sd">        as self.results, and these are parsed into two dictionaries,</span>
<span class="sd">        1) self.class_genome_bp, which holds the total number of annotated</span>
<span class="sd">        genomic bp for each class, and 2) self.class_sample_bp, which holds the</span>
<span class="sd">        total number of bp in the sample that overlapped each class.</span>

<span class="sd">        The table() method offers a means to filter/merge these</span>
<span class="sd">        fully-classified dictionaries such that they only contain featuretypes</span>
<span class="sd">        of interest (by using the `include` kwarg).</span>

<span class="sd">        The print_table() method prints a nice, optionally-filtered/merged</span>
<span class="sd">        table of the results, sorted by total sample bp.</span>

<span class="sd">        Example usage::</span>

<span class="sd">            &gt;&gt;&gt; bam = pybedtools.example_bedtool(&#39;x.bam&#39;)</span>
<span class="sd">            &gt;&gt;&gt; bed = bam.bam_to_bed(split=True).sort()</span>
<span class="sd">            &gt;&gt;&gt; anno = pybedtools.example_filename(&#39;dm3-chr2L-5M.gff.gz&#39;)</span>
<span class="sd">            &gt;&gt;&gt; names, fns = MultiClassifier.split_annotations(anno)</span>

<span class="sd">            &gt;&gt;&gt; # Example method of making a name converter.</span>
<span class="sd">            &gt;&gt;&gt; # (note that the split `fns` have the form &#39;split_exon.gff&#39;)</span>
<span class="sd">            &gt;&gt;&gt; def f(x):</span>
<span class="sd">            ...     return x.split(&#39;_&#39;)[-1]</span>

<span class="sd">            &gt;&gt;&gt; c = MultiClassifier(bed, fns, names=f, genome=&#39;dm3&#39;)</span>
<span class="sd">            &gt;&gt;&gt; c.classify()</span>
<span class="sd">            &gt;&gt;&gt; inc = [&#39;exon&#39;, &#39;intron&#39;]</span>
<span class="sd">            &gt;&gt;&gt; table = c.print_table(include=inc)</span>
<span class="sd">            &gt;&gt;&gt; print table #doctest: +NORMALIZE_WHITESPACE</span>
<span class="sd">            class	sample	genome</span>
<span class="sd">            exon	361971	1188410</span>
<span class="sd">            unannotated	19103	19841604</span>
<span class="sd">            exon, intron	14250	177913</span>
<span class="sd">            intron	10657	1803617</span>

<span class="sd">            &gt;&gt;&gt; # Clean up the split files.</span>
<span class="sd">            &gt;&gt;&gt; for fn in fns:</span>
<span class="sd">            ...     os.unlink(fn)</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bed</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="n">bed</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span> <span class="o">=</span> <span class="n">sample_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">genome</span>

        <span class="c1"># If a list of annotation files was provided, then use them directly,</span>
        <span class="c1"># ensuring that they are BedTools objects</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">annotations</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">)):</span>
            <span class="n">annotations</span> <span class="o">=</span> <span class="p">[</span><span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">.</span><span class="n">fn</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">annotations</span><span class="p">]</span>

            <span class="c1"># ... name-munge as necessary</span>
            <span class="k">if</span> <span class="n">names</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
                <span class="n">names</span> <span class="o">=</span> <span class="n">annotations</span>
            <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">names</span><span class="p">,</span> <span class="s1">&#39;__call__&#39;</span><span class="p">):</span>
                <span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="n">names</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">annotations</span><span class="p">]</span>

        <span class="c1"># Otherwise, split the single annotation file into a form suitible for</span>
        <span class="c1"># use by multiintersect</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">names</span><span class="p">,</span> <span class="n">files</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">split_annotations</span><span class="p">(</span><span class="n">annotations</span><span class="p">)</span>
            <span class="n">annotations</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">files</span><span class="p">)</span>
            <span class="n">names</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">names</span><span class="p">)</span>

        <span class="c1"># multiintersect still needs the input file (e.g., peaks)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">fn</span><span class="p">]</span> <span class="o">+</span> <span class="n">annotations</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">]</span> <span class="o">+</span> <span class="n">names</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">class_sample_bp</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">class_genome_bp</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">split_annotations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">annotations</span><span class="p">,</span> <span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;split_&#39;</span><span class="p">,</span> <span class="n">name_func</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given an annotation file in GFF format, split into different files --</span>
<span class="sd">        each file containing only one type of feature.</span>

<span class="sd">        `prefix` will be added to each featuretype to construct the filenames.</span>

<span class="sd">        `name_func`, by default, will use the feature type field of a GTF/GFF</span>
<span class="sd">        file, or the feature.name attribute of another format.  Supply a custom</span>
<span class="sd">        function that accepts a pybedtools.Interval instance for more control.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># dict of open files, one for each featuretype found</span>
        <span class="n">files</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">annotations</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="n">annotations</span><span class="p">)</span>
        <span class="n">ft</span> <span class="o">=</span> <span class="n">annotations</span><span class="o">.</span><span class="n">file_type</span>
        <span class="k">if</span> <span class="n">name_func</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">ft</span> <span class="o">==</span> <span class="s1">&#39;gff&#39;</span><span class="p">:</span>
                <span class="n">name_func</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">name_func</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">name</span>
        <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">annotations</span><span class="p">:</span>
            <span class="n">featuretype</span> <span class="o">=</span> <span class="n">name_func</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">featuretype</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
                <span class="n">filename</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">%s%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">prefix</span><span class="p">,</span> <span class="n">featuretype</span><span class="p">)</span>
                <span class="n">files</span><span class="p">[</span><span class="n">featuretype</span><span class="p">]</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span>
            <span class="n">files</span><span class="p">[</span><span class="n">featuretype</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">feature</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">files</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>
            <span class="n">f</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="p">[(</span><span class="n">featuretype</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                   <span class="k">for</span> <span class="p">(</span><span class="n">featuretype</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">files</span><span class="o">.</span><span class="n">items</span><span class="p">())]))</span>

    <span class="k">def</span> <span class="nf">classify</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Classify the features in self.bed, populating several dictionaries in</span>
<span class="sd">        self.  `kwargs` are passed on to BedTool.multi_intersect.  The</span>
<span class="sd">        &quot;empty=True&quot; kwarg to multi_intersect is always provided to make sure</span>
<span class="sd">        the classification works correctly.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">results</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">multi_intersect</span><span class="p">(</span>
            <span class="n">i</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="p">,</span>
            <span class="n">names</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">names</span><span class="p">,</span>
            <span class="n">genome</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span>
            <span class="n">empty</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

        <span class="n">sample</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">])</span>

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">results</span><span class="p">:</span>
            <span class="c1"># Even if there were multiple annotations, only report *that* there</span>
            <span class="c1"># was a hit, e.g., &quot;exon,exon,exon,5&#39;UTR&quot; -&gt; (exon, 5&#39;UTR)</span>
            <span class="n">full_class</span> <span class="o">=</span> <span class="nb">frozenset</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">))</span>

            <span class="c1"># Including sample name in class name would be redundant, so remove</span>
            <span class="c1"># it</span>
            <span class="n">class_name</span> <span class="o">=</span> <span class="n">full_class</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>

            <span class="c1"># Only report if sample was in the class</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span> <span class="ow">in</span> <span class="n">full_class</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">class_sample_bp</span><span class="p">[</span><span class="n">class_name</span><span class="p">]</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

            <span class="c1"># But always report the presence of the class, regardless of if</span>
            <span class="c1"># there was a hit in the sample or not.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">class_genome_bp</span><span class="p">[</span><span class="n">class_name</span><span class="p">]</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

        <span class="c1"># Genomic unannotated has the key [&quot;none&quot;]; sample unannotated as the</span>
        <span class="c1"># key [] (because it was set-differenced out)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">class_genome_bp</span><span class="p">[</span><span class="nb">frozenset</span><span class="p">([</span><span class="s1">&#39;unannotated&#39;</span><span class="p">])]</span> <span class="o">=</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">class_genome_bp</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="nb">frozenset</span><span class="p">([</span><span class="s1">&#39;none&#39;</span><span class="p">]),</span> <span class="mi">0</span><span class="p">)</span> \
            <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">class_genome_bp</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="nb">frozenset</span><span class="p">([]),</span> <span class="mi">0</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">class_sample_bp</span><span class="p">[</span><span class="nb">frozenset</span><span class="p">([</span><span class="s1">&#39;unannotated&#39;</span><span class="p">])]</span> <span class="o">=</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">class_sample_bp</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="nb">frozenset</span><span class="p">([]),</span> <span class="mi">0</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">include</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        If `include` is not None, then return versions of self.class_genome_bp</span>
<span class="sd">        and self.class_sample_bp that only look at the featuretypes in</span>
<span class="sd">        `include`.</span>

<span class="sd">        Otherwise, simply return these dictionaries unchanged (and including</span>
<span class="sd">        all available featuretypes)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">include</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">class_sample_bp</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">class_genome_bp</span>

        <span class="n">d</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
        <span class="n">s</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
        <span class="n">include</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">include</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">class_genome_bp</span><span class="p">:</span>
            <span class="c1"># orig data</span>
            <span class="n">seq_bp</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">class_genome_bp</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
            <span class="n">bed_bp</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">class_sample_bp</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>

            <span class="c1"># create a merged class name by keeping only featuretypes in</span>
            <span class="c1"># `include`</span>
            <span class="n">merged_class</span> <span class="o">=</span> <span class="nb">frozenset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">key</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">include</span><span class="p">))</span>

            <span class="c1"># update the new dictionaries</span>
            <span class="n">d</span><span class="p">[</span><span class="n">merged_class</span><span class="p">]</span> <span class="o">+=</span> <span class="n">bed_bp</span>
            <span class="n">s</span><span class="p">[</span><span class="n">merged_class</span><span class="p">]</span> <span class="o">+=</span> <span class="n">seq_bp</span>

        <span class="k">return</span> <span class="n">d</span><span class="p">,</span> <span class="n">s</span>

    <span class="k">def</span> <span class="nf">hierarchical_table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">order</span><span class="p">,</span> <span class="n">include</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                           <span class="n">include_unannotated</span><span class="o">=</span><span class="bp">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a hierarchically-ordered table, using the specified `order`.</span>

<span class="sd">        For example the order::</span>

<span class="sd">                [&#39;TSS&#39;, &#39;five_prime_UTR&#39;, &#39;CDS&#39;, &#39;intron&#39;, &#39;three_prime_UTR&#39;,</span>
<span class="sd">                &#39;TTS&#39;]</span>

<span class="sd">        would weight the classes from the 5&#39;-most end of the gene.</span>

<span class="sd">        This summarizes the classes based on the highest-priority featuretype</span>
<span class="sd">        in the hierarchy.</span>

<span class="sd">        For example, using the above hierarchy, the following summary-class</span>
<span class="sd">        assignments will be made::</span>

<span class="sd">            (TSS, five_prime_UTR)          -&gt; TSS</span>
<span class="sd">            (intron, CDS, TSS)             -&gt; TSS</span>
<span class="sd">            (intron, CDS, three_prime_UTR) -&gt; CDS</span>

<span class="sd">        The table has the following format, where the &quot;classes&quot; list is ordered</span>
<span class="sd">        by sample bp.</span>

<span class="sd">            {</span>
<span class="sd">                &#39;TSS&#39;: [</span>
<span class="sd">                        (&lt;class name 1&gt;, &lt;sample bp&gt;, &lt;genomic bp&gt;),</span>
<span class="sd">                        (&lt;class name 2&gt;, &lt;sample bp&gt;, &lt;genomic bp&gt;),</span>
<span class="sd">                        ...</span>
<span class="sd">                        ],</span>

<span class="sd">                &#39;five_prime_UTR&#39;: [</span>
<span class="sd">                        (&lt;class name 1&gt;, &lt;sample bp&gt;, &lt;genomic bp&gt;),</span>
<span class="sd">                        (&lt;class name 2&gt;, &lt;sample bp&gt;, &lt;genomic bp&gt;),</span>
<span class="sd">                        ...</span>
<span class="sd">                        ],</span>
<span class="sd">                ...</span>

<span class="sd">            }</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sample</span><span class="p">,</span> <span class="n">genomic</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">include</span><span class="o">=</span><span class="n">include</span><span class="p">)</span>
        <span class="n">sample</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">genomic</span> <span class="o">=</span> <span class="n">genomic</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">keys</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">genomic</span><span class="o">.</span><span class="n">keys</span><span class="p">())))</span>
        <span class="n">table</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">order</span><span class="p">:</span>
            <span class="n">classes</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">keys</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">k</span><span class="p">:</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">sample_bp</span> <span class="o">=</span> <span class="n">sample</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
                        <span class="n">genomic_bp</span> <span class="o">=</span> <span class="n">genomic</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
                        <span class="n">classes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                            <span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">sample_bp</span><span class="p">,</span> <span class="n">genomic_bp</span><span class="p">)</span>
                        <span class="p">)</span>
                    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                        <span class="c1"># already popped</span>
                        <span class="k">continue</span>
            <span class="n">table</span><span class="p">[</span><span class="n">h</span><span class="p">]</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">classes</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">include_unannotated</span><span class="p">:</span>
            <span class="n">table</span><span class="p">[</span><span class="s1">&#39;unannotated&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[(</span><span class="nb">frozenset</span><span class="p">([</span><span class="s1">&#39;unannotated&#39;</span><span class="p">]),</span>
                                     <span class="nb">sum</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span>
                                     <span class="nb">sum</span><span class="p">(</span><span class="n">genomic</span><span class="o">.</span><span class="n">values</span><span class="p">()))]</span>
        <span class="k">return</span> <span class="n">table</span>

    <span class="k">def</span> <span class="nf">print_table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">include</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a string containing a tab-delimited table, including header,</span>
<span class="sd">        with classes sorted by total bp in each class.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">d</span><span class="p">,</span> <span class="n">s</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">include</span><span class="p">)</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">out</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;class</span><span class="se">\t</span><span class="s1">sample</span><span class="se">\t</span><span class="s1">genome&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">cls</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">d</span><span class="p">[</span><span class="n">x</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="bp">True</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cls</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;unannotated&#39;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">cls</span><span class="p">))</span>

            <span class="n">out</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="n">d</span><span class="p">[</span><span class="n">cls</span><span class="p">],</span> <span class="n">s</span><span class="p">[</span><span class="n">cls</span><span class="p">]))</span>
        <span class="k">return</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>

<span class="n">MultiClassifier</span> <span class="o">=</span> <span class="n">BasePairClassifier</span>


<span class="k">class</span> <span class="nc">Classifier</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Classify intervals in one file by featuretypes in another.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bed</span><span class="p">,</span> <span class="n">annotations</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Classify features in `bed` -- typically a BED or SAM/BAM but can be any</span>
<span class="sd">        format supported by BedTools -- into classes based on the featuretypes</span>
<span class="sd">        in the GFF/GTF file, `annotations`.</span>

<span class="sd">        For example, you can classify ChIP-seq peaks in a BED file by intron,</span>
<span class="sd">        exon, or whatever is annotated in the GFF file.  If you want to</span>
<span class="sd">        consider promoter regions, you&#39;ll have to add these features yourself.</span>

<span class="sd">        The `class_counts` dictionary has its keys as sets of featuretypes</span>
<span class="sd">        (each one can be considered a &quot;class&quot; of features) and the value is the</span>
<span class="sd">        number of features in that class.  The special empty set class contains</span>
<span class="sd">        features that did not fall in an annotated region.</span>

<span class="sd">        You can access the individual features in the `class_features`</span>
<span class="sd">        dictionary, which contains the same keys but instead of counts, it</span>
<span class="sd">        contains the features themselves.  This is nice for saving the features</span>
<span class="sd">        in a separate BED file, e.g.,</span>

<span class="sd">        Furthermore, you can look up the class of any feature in the original</span>
<span class="sd">        BED file using the `feature_classes` dictionary::</span>


<span class="sd">        Example usage::</span>

<span class="sd">            &gt;&gt;&gt; bed = pybedtools.example_filename(&#39;gdc.bed&#39;)</span>
<span class="sd">            &gt;&gt;&gt; gff = pybedtools.example_filename(&#39;gdc.gff&#39;)</span>
<span class="sd">            &gt;&gt;&gt; c = pybedtools.contrib.classifier.Classifier(bed, gff)</span>
<span class="sd">            &gt;&gt;&gt; c.classify(include=[&#39;intron&#39;, &#39;exon&#39;])</span>
<span class="sd">            &gt;&gt;&gt; results = c.class_counts</span>
<span class="sd">            &gt;&gt;&gt; results == {</span>
<span class="sd">            ... frozenset([]): 1,</span>
<span class="sd">            ... frozenset([&#39;exon&#39;]): 3,</span>
<span class="sd">            ... frozenset([&#39;intron&#39;]): 3,</span>
<span class="sd">            ... frozenset([&#39;intron&#39;, &#39;exon&#39;]): 1}</span>
<span class="sd">            True</span>
<span class="sd">            &gt;&gt;&gt; key = frozenset([&#39;intron&#39;])</span>
<span class="sd">            &gt;&gt;&gt; features = c.class_features[key]</span>
<span class="sd">            &gt;&gt;&gt; pybedtools.BedTool(iter(features)).saveas() #doctest: +ELLIPSIS</span>
<span class="sd">            &lt;BedTool(...)&gt;</span>
<span class="sd">            &gt;&gt;&gt; feature = pybedtools.BedTool(bed)[2]</span>
<span class="sd">            &gt;&gt;&gt; c.feature_classes[feature] == set([&#39;intron&#39;, &#39;.&#39;])</span>
<span class="sd">            True</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bed</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="n">bed</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="n">annotations</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">file_type</span> <span class="o">!=</span> <span class="s1">&#39;gff&#39;</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Annotations file must be a GFF or GTF file; &#39;</span>
                             <span class="s1">&#39;</span><span class="si">%s</span><span class="s1"> appears to be a </span><span class="si">%s</span><span class="s1"> file&#39;</span> <span class="o">%</span> <span class="p">(</span>
                                 <span class="n">annotations</span><span class="p">,</span>
                                 <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">file_type</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">available_featuretypes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        List the featuretypes available in the annotations file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">featuretypes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="p">:</span>
            <span class="n">featuretypes</span><span class="o">.</span><span class="n">update</span><span class="p">([</span><span class="n">i</span><span class="p">[</span><span class="mi">2</span><span class="p">]])</span>
        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="n">featuretypes</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">classify</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">include</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">exclude</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">stranded</span><span class="o">=</span><span class="bp">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Perform classification, populating dictionaries in `self`.</span>

<span class="sd">        Intersect the BED file with the annotations file and return</span>
<span class="sd">        a dictionary where keys are BED features and values are the set of</span>
<span class="sd">        featuretypes that BED feature was found in.</span>


<span class="sd">        `include` is an optional list of featuretypes to restrict the</span>
<span class="sd">        classification to</span>

<span class="sd">        `exclude` is an optional list of featuretypes to exclude from</span>
<span class="sd">        classification (all other featuretypes will be used).</span>

<span class="sd">        To see what&#39;s available, use available_featuretypes().</span>

<span class="sd">        When run, this method creates the following dictionaries as attributes</span>
<span class="sd">        of this object:</span>

<span class="sd">         :feature_classes:</span>
<span class="sd">            keys are Intervals from `bed`; values are sets of featuretypes from</span>
<span class="sd">            `annotations`</span>

<span class="sd">         :class_features:</span>
<span class="sd">            keys are frozensets of featuretypes from `annotations`; values are</span>
<span class="sd">            lists of Intervals from `bed`;</span>

<span class="sd">         :class_counts:</span>
<span class="sd">            keys are frozensets of featuretypes from annotations`; values are</span>
<span class="sd">            number of features -- so the length of values in the class_features</span>
<span class="sd">            dictionary.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">include</span> <span class="ow">and</span> <span class="n">exclude</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Can only specify one of `include` or `exclude`&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">exclude</span><span class="p">:</span>
            <span class="n">exclude</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">exclude</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">include</span><span class="p">:</span>
            <span class="n">include</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">include</span><span class="p">)</span>

        <span class="c1"># Figure out the index of the featuretype field in the output</span>
        <span class="n">bed_fields</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">field_count</span><span class="p">()</span>
        <span class="n">featuretype_idx</span> <span class="o">=</span> <span class="n">bed_fields</span> <span class="o">+</span> <span class="mi">2</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">feature_classes</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>

        <span class="n">x</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">intersect</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="p">,</span>
                               <span class="n">wao</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                               <span class="n">s</span><span class="o">=</span><span class="n">stranded</span><span class="p">,</span>
                               <span class="n">stream</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">x</span><span class="p">:</span>
            <span class="n">featuretype</span> <span class="o">=</span> <span class="n">feature</span><span class="p">[</span><span class="n">featuretype_idx</span><span class="p">]</span>

            <span class="c1"># If we&#39;re not supposed to consider this featuretype, then reset to</span>
            <span class="c1"># the standard GFF empty field string of &quot;.&quot;</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">include</span> <span class="ow">and</span> <span class="n">featuretype</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">include</span><span class="p">)</span> \
                    <span class="ow">or</span> <span class="p">(</span><span class="n">exclude</span> <span class="ow">and</span> <span class="n">featuretype</span> <span class="ow">in</span> <span class="n">exclude</span><span class="p">):</span>
                <span class="n">featuretype</span> <span class="o">=</span> <span class="s1">&#39;.&#39;</span>

            <span class="c1"># the original query is in the first `bed_fields` items.  Construct</span>
            <span class="c1"># a new feature out of this and use it as the key.</span>
            <span class="n">key</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">create_interval_from_list</span><span class="p">(</span>
                <span class="n">feature</span><span class="o">.</span><span class="n">fields</span><span class="p">[:</span><span class="n">bed_fields</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">feature_classes</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">([</span><span class="n">featuretype</span><span class="p">])</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">class_features</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">class_counts</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">feature</span><span class="p">,</span> <span class="n">featuretypes</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">feature_classes</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
            <span class="c1"># get rid of &quot;unannotated&quot;</span>
            <span class="n">ft</span> <span class="o">=</span> <span class="n">featuretypes</span><span class="o">.</span><span class="n">difference</span><span class="p">([</span><span class="s1">&#39;.&#39;</span><span class="p">])</span>
            <span class="n">key</span> <span class="o">=</span> <span class="nb">frozenset</span><span class="p">(</span><span class="n">ft</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">class_features</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">class_counts</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="c1"># convert defaultdicts to regular dicts</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">class_features</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">class_features</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">class_counts</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">class_counts</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">feature_classes</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">feature_classes</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">features_to_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Writes a set of files, one for each class.</span>

<span class="sd">        The filenames will be constructed based on the class names, track lines</span>
<span class="sd">        will be added to indicate classes, and `prefix` and `suffix` will be</span>
<span class="sd">        added to the filenames.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">def</span> <span class="nf">make_filename</span><span class="p">(</span><span class="n">klass</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">klass</span><span class="p">)))</span> <span class="o">+</span> <span class="n">suffix</span>

        <span class="k">def</span> <span class="nf">make_trackline</span><span class="p">(</span><span class="n">klass</span><span class="p">):</span>
            <span class="k">return</span> <span class="s1">&#39;track name=&quot;</span><span class="si">%s</span><span class="s1">&quot;&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">klass</span><span class="p">))))</span>

        <span class="k">for</span> <span class="n">klass</span><span class="p">,</span> <span class="n">features</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">class_features</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="n">features</span><span class="p">)</span>\
                <span class="o">.</span><span class="n">saveas</span><span class="p">(</span><span class="n">make_filename</span><span class="p">(</span><span class="n">klass</span><span class="p">),</span> <span class="n">make_trackline</span><span class="p">(</span><span class="n">klass</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">hierarchical_table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">order</span><span class="p">,</span> <span class="n">include</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                           <span class="n">include_unannotated</span><span class="o">=</span><span class="bp">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a hierarchically-ordered table, using the specified `order`.</span>

<span class="sd">        For example the order::</span>

<span class="sd">            [&#39;TSS&#39;, &#39;five_prime_UTR&#39;, &#39;CDS&#39;, &#39;intron&#39;, &#39;three_prime_UTR&#39;,</span>
<span class="sd">            &#39;TTS&#39;]</span>

<span class="sd">        would weight the classes from the 5&#39;-most end of the gene.</span>

<span class="sd">        This summarizes the classes based on the highest-priority featuretype</span>
<span class="sd">        in the hierarchy.</span>

<span class="sd">        For example, using the above hierarchy, the following summary-class</span>
<span class="sd">        assignments will be made::</span>

<span class="sd">            (TSS, five_prime_UTR)          -&gt; TSS</span>
<span class="sd">            (intron, CDS, TSS)             -&gt; TSS</span>
<span class="sd">            (intron, CDS, three_prime_UTR) -&gt; CDS</span>

<span class="sd">        The table has the following format, where the &quot;classes&quot; list is ordered</span>
<span class="sd">        by sample bp.</span>

<span class="sd">            {</span>
<span class="sd">                &#39;TSS&#39;: [</span>
<span class="sd">                        (&lt;class name 1&gt;, count),</span>
<span class="sd">                        (&lt;class name 2&gt;, count),</span>
<span class="sd">                        ...</span>
<span class="sd">                        ],</span>

<span class="sd">                &#39;five_prime_UTR&#39;: [</span>
<span class="sd">                        (&lt;class name 1&gt;, count),</span>
<span class="sd">                        (&lt;class name 2&gt;, count),</span>
<span class="sd">                        ...</span>
<span class="sd">                        ],</span>
<span class="sd">                ...</span>

<span class="sd">            }</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">counts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">include</span><span class="o">=</span><span class="n">include</span><span class="p">)</span>
        <span class="n">counts</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">keys</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">counts</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="n">table</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">order</span><span class="p">:</span>
            <span class="n">classes</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">keys</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">k</span><span class="p">:</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">count</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
                        <span class="n">classes</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">k</span><span class="p">,</span> <span class="n">count</span><span class="p">))</span>
                    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                        <span class="c1"># i.e., already popped off</span>
                        <span class="k">continue</span>
            <span class="n">table</span><span class="p">[</span><span class="n">h</span><span class="p">]</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">classes</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">include_unannotated</span><span class="p">:</span>
            <span class="n">table</span><span class="p">[</span><span class="s1">&#39;unannotated&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[(</span><span class="nb">frozenset</span><span class="p">([</span><span class="s1">&#39;unannotated&#39;</span><span class="p">]),</span> <span class="nb">sum</span><span class="p">(</span><span class="n">counts</span><span class="o">.</span><span class="n">values</span><span class="p">()))]</span>
        <span class="k">return</span> <span class="n">table</span>

    <span class="k">def</span> <span class="nf">table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">include</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        If `include` is not None, then return a copy of self.class_counts that</span>
<span class="sd">        only have at the featuretypes in `include`.</span>

<span class="sd">        Otherwise, return a simple copy.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">include</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">class_counts</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="n">d</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
        <span class="n">s</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
        <span class="n">include</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">include</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">class_counts</span><span class="p">:</span>
            <span class="c1"># orig data</span>
            <span class="n">c</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">class_counts</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>

            <span class="c1"># create a merged class name by keeping only featuretypes in</span>
            <span class="c1"># `include`.  If none of the members were in `include`, then the</span>
            <span class="c1"># counts are appended to entry keyed by set().</span>
            <span class="n">merged_class</span> <span class="o">=</span> <span class="nb">frozenset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">key</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">include</span><span class="p">))</span>

            <span class="c1"># update the new dictionaries</span>
            <span class="n">d</span><span class="p">[</span><span class="n">merged_class</span><span class="p">]</span> <span class="o">+=</span> <span class="n">c</span>

        <span class="k">return</span> <span class="n">d</span>


    <span class="k">def</span> <span class="nf">pie</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">hierarchical_table</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        `hierarchical_table` is the result of calling self.hierarchical_table()`.</span>

<span class="sd">        `ax` is an optional Axes object to plot onto, otherwise a new figure and axes will be created.</span>

<span class="sd">        `order`, if None, will sort the items from largest to smallest.</span>
<span class="sd">        Otherwise, `order` is the order of keys in `hierarchical_table` in</span>
<span class="sd">        which they should be plotted, counter-clockwise from the positive</span>
<span class="sd">        x axis.</span>

<span class="sd">        Additional args are passed to ax.pie.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>

        <span class="c1"># restructure the table so that it&#39;s a dict of class sums (rather than</span>
        <span class="c1"># having every class)</span>
        <span class="n">d</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">hierarchical_table</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
            <span class="n">d</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">v</span><span class="p">)</span>

        <span class="n">total</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
        <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
            <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">111</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">order</span><span class="p">:</span>
            <span class="n">items</span> <span class="o">=</span> <span class="p">[(</span><span class="n">k</span><span class="p">,</span> <span class="n">d</span><span class="p">[</span><span class="n">k</span><span class="p">])</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">order</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">items</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">items</span><span class="p">()),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

        <span class="n">newlabels</span><span class="p">,</span> <span class="n">labels</span><span class="p">,</span> <span class="n">counts</span> <span class="o">=</span> <span class="p">[],</span> <span class="p">[],</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">label</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">items</span><span class="p">:</span>
            <span class="n">labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">label</span><span class="p">)</span>
            <span class="n">counts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">count</span><span class="p">)</span>
            <span class="n">frac</span> <span class="o">=</span> <span class="n">count</span> <span class="o">/</span> <span class="n">total</span> <span class="o">*</span> <span class="mf">100.</span>
            <span class="n">newlabels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;{label}: {count} ({frac:.1f}%)&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="o">**</span><span class="nb">locals</span><span class="p">()))</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">pie</span><span class="p">(</span>
            <span class="n">x</span><span class="o">=</span><span class="n">counts</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">newlabels</span><span class="p">,</span> <span class="n">labeldistance</span><span class="o">=</span><span class="mf">1.2</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">ax</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  <li><a href="../../index.html">Module code</a><ul>
  <li><a href="../../pybedtools.html">pybedtools</a><ul>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>

    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">pybedtools 0.7.9 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="../../pybedtools.html" accesskey="U">pybedtools</a> &#187;</li> 
      </ul>
    </div>


    <div class="footer">
      &copy;2010-2015, Ryan Dale.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.5.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.8</a>
      
    </div>

    

    
  </body>
</html>