<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>pybedtools overview and examples &mdash; pybedtools v0.1 documentation</title>
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '0.1',
        COLLAPSE_MODINDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="pybedtools v0.1 documentation" href="index.html" />
    <link rel="next" title="Class documentation" href="autodocs.html" />
    <link rel="prev" title="Welcome to pybedtools’s documentation!" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="autodocs.html" title="Class documentation"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to pybedtools’s documentation!"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">pybedtools v0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="pybedtools-overview-and-examples">
<h1><tt class="docutils literal"><span class="pre">pybedtools</span></tt> overview and examples<a class="headerlink" href="#pybedtools-overview-and-examples" title="Permalink to this headline">¶</a></h1>
<p>Python wrapper for Aaron Quinlan&#8217;s <tt class="docutils literal"><span class="pre">BEDtools</span></tt> (<a class="reference" href="http://code.google.com/p/bedtools/">http://code.google.com/p/bedtools/</a>).</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">See <tt class="docutils literal"><span class="pre">doc/_build/html/index.html</span></tt> in the source distribution for
complete documentation of all the <tt class="docutils literal"><span class="pre">bedtool</span></tt> methods.</p>
</div>
<p>The <tt class="docutils literal"><span class="pre">bedtool</span></tt> object has methods that call the tools in the BEDtools suite.
Where appropriate, the method returns another <tt class="docutils literal"><span class="pre">bedtool</span></tt> object, so you can
chain or pipe things together just like from the command line.  There are also
other features not [yet] found in BEDtools like random selection of features,
obtaining the centers of features (e.g., for motif detection), reciprocal
intersection reports, feature counts, permutation testing, and much more.</p>
<div class="section" id="general-usage">
<h2>General usage<a class="headerlink" href="#general-usage" title="Permalink to this headline">¶</a></h2>
<p>Many methods in a <tt class="docutils literal"><span class="pre">bedtool</span></tt> object make system calls to various
programs in the <tt class="docutils literal"><span class="pre">BEDtools</span></tt> suite.  In most cases the return value is a
new <tt class="xref docutils literal"><span class="pre">bedtool</span></tt> object.  This allows you to chain together analyses
just like piping commands together at the command line.</p>
<p>The methods in this class mimic the command line usage of <tt class="docutils literal"><span class="pre">BEDtools</span></tt> as
closely as possible.  Any flag that can be passed to the <tt class="docutils literal"><span class="pre">BEDtools</span></tt>
programs can be passed as a keyword argument (see examples below).
Typically, for convenience <tt class="docutils literal"><span class="pre">-i</span></tt>, <tt class="docutils literal"><span class="pre">-a</span></tt>, and <tt class="docutils literal"><span class="pre">-b</span></tt> are already passed
for you although you can override this by passing these keyword arguments
explicitly. <tt class="docutils literal"><span class="pre">BEDtools</span></tt> flags that require an argument, for example the
window size <tt class="docutils literal"><span class="pre">-w</span></tt> in <tt class="docutils literal"><span class="pre">windowBed</span></tt>, are simply passed as integers.
Boolean flags (e.g., <tt class="docutils literal"><span class="pre">-sw</span></tt> in <tt class="docutils literal"><span class="pre">windowBed</span></tt>) are passed as Python
booleans.</p>
<p>Typical workflow is to set up a bedtool object with a bed file you already have:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">a</span> <span class="o">=</span> <span class="n">bedtool</span><span class="p">(</span><span class="s">&#39;in.bed&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p><cite>a</cite> now references the file <tt class="docutils literal"><span class="pre">in.bed</span></tt> on disk.</p>
<p>Using BEDtools from the command line, in order to get a new bed file of the
intersection of this file with another bed file, <tt class="docutils literal"><span class="pre">other.bed</span></tt> only
returning uniquely intersecting features from <tt class="docutils literal"><span class="pre">in.bed</span></tt>  we might use:</p>
<div class="highlight-python"><pre>intersectBed -a in.bed -b other.bed -u &gt; intersection.bed</pre>
</div>
<p>Using <tt class="docutils literal"><span class="pre">pybedtools</span></tt>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">b</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">intersect</span><span class="p">(</span><span class="s">&#39;other.bed&#39;</span><span class="p">,</span> <span class="n">u</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<p>This creates a new temp file in <tt class="docutils literal"><span class="pre">/tmp</span></tt> by default, but you can change where
temp files are saved using <tt class="docutils literal"><span class="pre">pybedtools.set_tempdir()</span></tt>.  To save a file
explicitly and add a trackline that will label it in a genome browser, use <tt class="docutils literal"><span class="pre">saveas()</span></tt>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">b</span><span class="o">.</span><span class="n">saveas</span><span class="p">(</span><span class="s">&#39;intersection.bed&#39;</span><span class="p">,</span><span class="n">trackline</span><span class="o">=</span><span class="s">&#39;track name=&quot;intersection&quot;&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Most methods return new <tt class="docutils literal"><span class="pre">bedtool</span></tt> objects, allowing you to chain things together.  To give you
a flavor of this, here&#8217;s how you would get 10 random centers of features that are unique to the file
<tt class="docutils literal"><span class="pre">other.bed</span></tt>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">b</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">intersect</span><span class="p">(</span><span class="s">&#39;other.bed&#39;</span><span class="p">,</span><span class="n">v</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span><span class="o">.</span><span class="n">feature_centers</span><span class="p">(</span><span class="mf">100</span><span class="p">)</span><span class="o">.</span><span class="n">random_subset</span><span class="p">(</span><span class="mf">10</span><span class="p">)</span>
</pre></div>
</div>
<p>When you&#8217;re done, it&#8217;s a good idea to clean up temporary files.  Temp files
are never deleted until you explicitly say so:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">pybedtools</span><span class="o">.</span><span class="n">cleanup</span><span class="p">()</span>
</pre></div>
</div>
<p>If you forget to call <tt class="docutils literal"><span class="pre">pybedtools.cleanup()</span></tt>, you can always manually delete
the files from your temp dir (typically <tt class="docutils literal"><span class="pre">/tmp</span></tt>).  They are the files that
follow the pattern <tt class="docutils literal"><span class="pre">pybedtools.*.tmp</span></tt>.</p>
<p>The keyword arguments to methods are passed directly to the BEDtools
programs.  On/off switches (e.g., <tt class="docutils literal"><span class="pre">-c</span></tt>, <tt class="docutils literal"><span class="pre">-u</span></tt>, or <tt class="docutils literal"><span class="pre">-v</span></tt> for <tt class="docutils literal"><span class="pre">intersectBed</span></tt>) are
called with a boolean kwarg; others (like <tt class="docutils literal"><span class="pre">-f</span></tt> for <tt class="docutils literal"><span class="pre">intersectBed</span></tt>) are
passed in like a normal kwarg.  For example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">a</span> <span class="o">=</span> <span class="n">bedtool</span><span class="p">(</span><span class="s">&#39;in.bed&#39;</span><span class="p">)</span>
<span class="n">a</span><span class="o">.</span><span class="n">intersect</span><span class="p">(</span><span class="s">&#39;other.bed&#39;</span><span class="p">,</span> <span class="n">v</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">f</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
</pre></div>
</div>
<p>Importantly, the docstring for a method automatically includes the help text of
the original <tt class="docutils literal"><span class="pre">BEDtools</span></tt> program, so you can check which kwargs you can use
directly from the interpreter.</p>
<div class="section" id="examples">
<h3>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<div class="section" id="example-flanking-seqs">
<h2>Example: Flanking seqs<a class="headerlink" href="#example-flanking-seqs" title="Permalink to this headline">¶</a></h2>
<p>The <tt class="docutils literal"><span class="pre">slop()</span></tt> method (which calls <tt class="docutils literal"><span class="pre">slopBed</span></tt>) needs a chromosome size
file.  If you specify a genome name to the <tt class="docutils literal"><span class="pre">slop()</span></tt> method, it will
retrieve this file for you automatically from the UCSC Genome Browser MySQL
database.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">pybedtools</span>
<span class="n">a</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">bedtool</span><span class="p">(</span><span class="s">&#39;in.bed&#39;</span><span class="p">)</span>
<span class="n">extended</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">slop</span><span class="p">(</span><span class="n">genome</span><span class="o">=</span><span class="s">&#39;dm3&#39;</span><span class="p">,</span><span class="n">l</span><span class="o">=</span><span class="mf">100</span><span class="p">,</span><span class="n">r</span><span class="o">=</span><span class="mf">100</span><span class="p">)</span>
<span class="n">flanking</span> <span class="o">=</span> <span class="n">extended</span><span class="o">.</span><span class="n">subtract</span><span class="p">(</span><span class="n">a</span><span class="p">)</span><span class="o">.</span><span class="n">saveas</span><span class="p">(</span><span class="s">&#39;flanking.bed&#39;</span><span class="p">)</span>
<span class="n">flanking</span><span class="o">.</span><span class="n">sequence</span><span class="p">(</span><span class="n">fi</span><span class="o">=</span><span class="s">&#39;dm3.fa&#39;</span><span class="p">)</span>
<span class="n">flanking</span><span class="o">.</span><span class="n">save_seqs</span><span class="p">(</span><span class="s">&#39;flanking.fa&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Or, as a one-liner:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">pybedtools</span><span class="o">.</span><span class="n">bedtool</span><span class="p">(</span><span class="s">&#39;in.bed&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">slop</span><span class="p">(</span><span class="n">genome</span><span class="o">=</span><span class="s">&#39;dm3&#39;</span><span class="p">,</span><span class="n">l</span><span class="o">=</span><span class="mf">100</span><span class="p">,</span><span class="n">r</span><span class="o">=</span><span class="mf">100</span><span class="p">)</span><span class="o">.</span><span class="n">subtract</span><span class="p">(</span><span class="n">a</span><span class="p">)</span><span class="o">.</span><span class="n">sequence</span><span class="p">(</span><span class="n">fi</span><span class="o">=</span><span class="s">&#39;dm3.fa&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">save_seqs</span><span class="p">(</span><span class="s">&#39;flanking.fa&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Don&#8217;t forget to clean up!:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">pybedtools</span><span class="o">.</span><span class="n">cleanup</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="section" id="example-region-centers-that-are-fully-intergenic">
<h2>Example: Region centers that are fully intergenic<a class="headerlink" href="#example-region-centers-that-are-fully-intergenic" title="Permalink to this headline">¶</a></h2>
<p>Useful for, e.g., motif searching:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">a</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">bedtool</span><span class="p">(</span><span class="s">&#39;in.bed&#39;</span><span class="p">)</span>

<span class="c"># Sort by score</span>
<span class="n">a</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">sorted</span><span class="p">(</span><span class="n">col</span><span class="o">=</span><span class="mf">5</span><span class="p">,</span><span class="n">reverse</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

<span class="c"># Exclude some regions</span>
<span class="n">a</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">subtract</span><span class="p">(</span><span class="s">&#39;regions-to-exclude.bed&#39;</span><span class="p">)</span>

<span class="c"># Get 100 bp on either side of center</span>
<span class="n">a</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">peak_centers</span><span class="p">(</span><span class="mf">100</span><span class="p">)</span><span class="o">.</span><span class="n">saveas</span><span class="p">(</span><span class="s">&#39;200-bp-peak-centers.bed&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="example-histogram-of-feature-lengths">
<h2>Example: Histogram of feature lengths<a class="headerlink" href="#example-histogram-of-feature-lengths" title="Permalink to this headline">¶</a></h2>
<p>Note that you need matplotlib installed to plot the histogram.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">pylab</span> <span class="kn">as</span> <span class="nn">p</span>
<span class="n">a</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">bedtool</span><span class="p">(</span><span class="s">&#39;in.bed&#39;</span><span class="p">)</span>
<span class="n">p</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">lengths</span><span class="p">(),</span><span class="n">bins</span><span class="o">=</span><span class="mf">50</span><span class="p">)</span>
<span class="n">p</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <h3><a href="index.html">Table Of Contents</a></h3>
            <ul>
<li><a class="reference" href=""><tt class="docutils literal"><span class="pre">pybedtools</span></tt> overview and examples</a><ul>
<li><a class="reference" href="#general-usage">General usage</a><ul>
<li><a class="reference" href="#examples">Examples</a></li>
</ul>
</li>
<li><a class="reference" href="#example-flanking-seqs">Example: Flanking seqs</a></li>
<li><a class="reference" href="#example-region-centers-that-are-fully-intergenic">Example: Region centers that are fully intergenic</a></li>
<li><a class="reference" href="#example-histogram-of-feature-lengths">Example: Histogram of feature lengths</a></li>
</ul>
</li>
</ul>

            <h4>Previous topic</h4>
            <p class="topless"><a href="index.html"
                                  title="previous chapter">Welcome to pybedtools&#8217;s documentation!</a></p>
            <h4>Next topic</h4>
            <p class="topless"><a href="autodocs.html"
                                  title="next chapter">Class documentation</a></p>
            <h3>This Page</h3>
            <ul class="this-page-menu">
              <li><a href="_sources/README.txt"
                     rel="nofollow">Show Source</a></li>
            </ul>
          <div id="searchbox" style="display: none">
            <h3>Quick search</h3>
              <form class="search" action="search.html" method="get">
                <input type="text" name="q" size="18" />
                <input type="submit" value="Go" />
                <input type="hidden" name="check_keywords" value="yes" />
                <input type="hidden" name="area" value="default" />
              </form>
              <p class="searchtip" style="font-size: 90%">
              Enter search terms or a module, class or function name.
              </p>
          </div>
          <script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="autodocs.html" title="Class documentation"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to pybedtools’s documentation!"
             >previous</a> |</li>
        <li><a href="index.html">pybedtools v0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
      &copy; Copyright 2010, Ryan Dale.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 0.6.3.
    </div>
  </body>
</html>